### PR TITLE
sysctl finetune and resource restrictions

### PR DESCRIPTION
**Pull request by Ecsy**
sysctl finetune and resource restrictions

### CODE REVIEW

File: cmd/pke/app/phases/runtime/kubernetes/kubelet_kernel_params.yaml.tmpl [hunk 1]
```diff
@@ -1,3 +1,6 @@
 vm.overcommit_memory=1
+# vm.overcommit_ratio=100
+vm.oom_kill_allocating_task=1
```
Comment (orymate) [inline cmd/pke/app/phases/runtime/kubernetes/kubelet_kernel_params.yaml.tmpl]: > If this is set to non-zero, the OOM killer simply kills the task that triggered the out-of-memory condition. This avoids the expensive tasklist scan.

Do we know that the tasklist scanning is too expensive? Or is it better to kill a random task that failed to allocate the last 1KB of memory instead of the one that caused the out of memory condition?

File: cmd/pke/app/phases/runtime/kubernetes/kubelet_kernel_params.yaml.tmpl [hunk 2]
```diff
@@ -1,3 +1,6 @@
...
 kernel.panic=10
 kernel.panic_on_oops=1
+vm.swappiness=0
```
Comment (orymate) [inline cmd/pke/app/phases/runtime/kubernetes/kubelet_kernel_params.yaml.tmpl]: I don't see a reason to set this. Normally we don't even have swap volumes, but if we do, it may be a smart decision to swap out pages that well be never accessed to gain place for fs cache. I would not change these settings without well documented or own measurements.

File: cmd/pke/app/phases/kubeadm/node/kubeadm.go [hunk 3]
```diff
@@ -70,7 +70,7 @@ func (n Node) writeKubeadmConfig(out io.Writer, filename string) error {
 	case 12, 13:
 		// see https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha3
 		conf = kubeadmConfigV1Alpha3Template()
-	case 14:
+	case 14, 15:
```
Comment (orymate) [inline cmd/pke/app/phases/kubeadm/node/kubeadm.go]: this could be in a separate commit.

File: cmd/pke/app/phases/runtime/kubernetes/kubelet_kernel_params.yaml.tmpl [hunk 4]
```diff
@@ -1,3 +1,6 @@
 vm.overcommit_memory=1
+# vm.overcommit_ratio=100
```
Comment (orymate) [inline cmd/pke/app/phases/runtime/kubernetes/kubelet_kernel_params.yaml.tmpl]: this would be only effective if overcommit_memory is 2 (with value 1, the overcommit_ratio is effectively infinite). why do we keep it here?

File: cmd/pke/app/phases/runtime/kubernetes/kubelet_kernel_params.yaml.tmpl (hunk 1 above)
Comment (Ecsy) [reply cmd/pke/app/phases/runtime/kubernetes/kubelet_kernel_params.yaml.tmpl]: Statement `one that caused the out of memory condition` does not stand. By default the Linux Kernel uses heuristics to determine which process gets killed or not. Setting `oom_score_adj` can lower the chance of being terminated, but that requires fine-tuning and it is based on PID. In general we have seen sshd or even Kuberentes CNI gets killed before the workload which caused the actual situation. The probability of a higher memory consumer process tries to allocate more memory is higher.

Comment (orymate) [reply cmd/pke/app/phases/runtime/kubernetes/kubelet_kernel_params.yaml.tmpl]: this can be true (i'm not sure if this is a good solution), but then at least document it in the code
